sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -679,6 +705,19 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
 
   // FloatConverter
   private[this] def castToFloat(from: DataType): Any => Any = from match {
+    case StringType if ansiEnabled =>
+      buildCast[UTF8String](_, s => {
+        val floatStr = s.toString
+        try floatStr.toFloat catch {
+          case _: NumberFormatException =>
+            val f = Cast.processFloatingPointSpecialLiterals(floatStr, true)
+            if (f == null) {
```

Comment:
this is too much code duplication. How about unifying these 2 cases?
```
val f = Cast.processFloatingPointSpecialLiterals(floatStr, true)
if (f == null && ansiEnabled) {
  throw ...
} else {
  f
}
```

Reply:
Done